File: scala/src/main/scala/CodemotionTask.scala
object CodemotionTask {

  object MealComponents {

    trait Ingredient

    trait MixedIngredient[F <: Ingredient, S <: Ingredient] extends Ingredient {
      def firstIngredient: F

      def secondIngredient: S
    }

    trait Spice extends Ingredient

    trait Meal

    trait SeasonedMeal extends Meal

  }

  object RecipeSteps {

    import MealComponents._

    trait CookingStep[I1 <: Ingredient, R <: Meal] {
      def cook(i: Seq[I1]): Seq[R]
    }

    trait MixingStep[I1 <: Ingredient, I2 <: Ingredient, R <: MixedIngredient[I1, I2]] {
      def mix(i1: Seq[I1], i2: Seq[I2]): Seq[R]
    }

    trait SeasoningStep[M <: Meal, I <: Spice, R <: SeasonedMeal] {
      def season(m: Seq[M], s: Seq[I]): Seq[R]
    }

  }

  object Recipes {

    import MealComponents._
    import RecipeSteps._

    trait Recipe[M <: SeasonedMeal] {
      def makeRecipe: Seq[M]
    }

    trait MixAndCookAndSeasonRecipe[M <: SeasonedMeal, I1 <: Ingredient, I2 <: Ingredient, MI <: MixedIngredient[I1, I2], N <: Meal, S <: Spice] extends Recipe[SeasonedMeal] {
      def i1: Seq[I1]

      def i2: Seq[I2]

      def s: Seq[S]

      def mixingStep: MixingStep[I1, I2, MI]

      def cookingStep: CookingStep[MI, N]

      def seasoningStep: SeasoningStep[N, S, M]

      override def makeRecipe: Seq[M] = {
        val mixed: Seq[MI] = mixingStep.mix(i1, i2)
        val cooked: Seq[N] = cookingStep.cook(mixed)
        val seasoned = seasoningStep.season(cooked, s)
        seasoned
      }
    }

  }

}

File: java/CodemotionTask.java
package com.lieferando;

abstract class Ingredient {}
abstract class Spice extends Ingredient {}
abstract class Meal {}
abstract class SeasonedMeal extends Meal {}
abstract class MixedIngredient<I1 extends Ingredient,
        I2 extends Ingredient> extends Ingredient {}

interface CookingStep<I extends Ingredient, R extends Meal> {
    R cook(I ingredient);
}

interface MixingStep<I1 extends Ingredient, I2 extends
        Ingredient, R extends MixedIngredient<I1, I2>> {
    R mix(I1 firstIng, I2 secondIng);
}

interface SeasoningStep<M extends Meal, S extends Spice,
        R extends SeasonedMeal> {
    R season(M meal, S spice);
}

abstract class Recipe<R extends SeasonedMeal> {
    abstract R doRecipe();
}

abstract class MixAndCookAndSeasonRecipe<R extends 
        SeasonedMeal, I1 extends Ingredient, I2 extends 
        Ingredient, MI extends MixedIngredient<I1, I2>, N 
        extends Meal, S extends Spice> extends Recipe<R> {

    I1 i1;
    I2 i2;
    S s;
    CookingStep<MI, N> cookingStep;
    MixingStep<I1, I2, MI> mixingStep;
    SeasoningStep<N, S, R> seasoningStep;

    @Override
    R doRecipe() {
        MI mixed = mixingStep.mix(i1, i2);
        N cooked = cookingStep.cook(mixed);
        R seasoned = seasoningStep.season(cooked, s);
        return seasoned;
    }
}
